### ISSUE TITLE
Training models when data doesnt fit in memory

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10792]
Training models when data doesn’t fit in memory<br>
<br>
https://ift.tt/2JYpAPr<br>
<br>
Guilherme Duarte Marmerola<br>
<br>

<div><div><div><div><article><p>At some point in your carreer in data science, you’ll deal with some big dataset which will bring chaos to your otherwise clean workflow: pandas will crash with a <code>MemoryError</code>, all of the models in <code>sklearn</code> will seem useless as they need all of the data in RAM, as well as the coolest new methods you started to use, like <code>UMAP</code> (what did you expect? That the author would create a cutting edge ML algorithm <em>and</em> a distributed, out-of-memory implementation?).</p>
<p>You believe that distributed ML training can solve your problem. You start to do some digging on Hadoop, Hive, Spark, Kubernetes, etc and learn that they really <em>could</em> help you scale your models. However, you also learn that configuring clusters of machines is really hard, and there’s not a single, unified abstraction (like <code>sklearn</code>) for distributed ML training.</p>
<p>Your deadline is approaching, you haven’t scratched the surface on distributed computing, and start to imagine if there’s a quicker and may I say, <em>dirtier</em> approach.</p>
<p>And there is! And it is actually not dirty at all! In this post, I’ll show some hacks to make your favorite model work in a single-node environment and a cool tool (<strong>Dask</strong>) that will let you grasp distributed computation in your cozy python environment and with your well known pandas syntax.</p>
<p>I hope that the tricks I’ll show will expand the array of problems you can solve with a single machine, help you understand a little bit more of distributed computing, and buy you time to study all the other tools!</p>
<h2>Tracking memory usage</h2>
<p>We will build several ML pipelines and track their memory usage to see our tricks at work. In order to do this, I created a decorator that uses the <code>tracemalloc</code> and <code>concurrent</code> packages to sample memory usage in a function every 1 ms and return its results, drawing inspiration from this <a href="https://medium.com/survata-engineering-blog/monitoring-memory-usage-of-a-running-python-program-49f027e3d1ba">post</a>. You can check the full code <a href="https://github.com/gdmarmerola/big-data-ml-training/blob/master/track_memory.py">here</a>.</p>
<div><div><pre><code><span># importing custom memory tracking decorator
</span><span>from</span> <span>track_memory</span> <span>import</span> <span>track_memory_use</span><span>,</span> <span>plot_memory_use</span>
</code></pre></div></div>
<h2>Data</h2>
<p>For this example, we’ll use the fraud detection dataset from <a href="https://www.kaggle.com/ntnu-testimon/paysim1">Synthetic Financial Datasets For Fraud Detection</a> on Kaggle, with some feature engineering tricks borrowed from the Kaggle Kernel <a href="https://www.kaggle.com/arjunjoshua/predicting-fraud-in-financial-payment-services">Predicting Fraud in Financial Payment Services</a>. The data is not huge and it actually fits in memory, but it’s big enough so we can demonstrate memory usage gains with our tricks.</p>
<p>We have two datasets: a big <code>train.csv</code> (80% of records) for model training and a small <code>test.csv</code> (20% of records) for model validation (for simplicity, we’ll only use one validation set). The data invites a binary classification task to predict which transactions are fraudulent, given the amount transacted, source and destination balances, and other information.</p>
<h2>Naive implementation: A Tale of Memory Usage</h2>
<p>Let us solve the fraud problem in a naive way and track memory usage. The first thing we usually do is to read the dataset with pandas. How much does this cost, memory-wise?</p>
<div><div><pre><code><span># importing
</span><span>import</span> <span>pandas</span> <span>as</span> <span>pd</span>

<span># using a function so we can track memory usage
</span><span>@</span><span>track_memory_use</span><span>(</span><span>close</span><span>=</span><span>False</span><span>,</span> <span>return_history</span><span>=</span><span>True</span><span>)</span>
<span>def</span> <span>pandas_read</span><span>():</span>
    
    <span># reading train data
</span>    <span>df_train</span> <span>=</span> <span>pd</span><span>.</span><span>read_csv</span><span>(</span><span>'./train.csv'</span><span>)</span>
    
    <span>return</span> <span>df_train</span>

<span># executing
</span><span>df_train</span><span>,</span> <span>mem_history_1</span> <span>=</span> <span>pandas_read</span><span>()</span>
</code></pre></div></div>
<div><div><pre><code>Current memory usage: 570.384648
Peak memory usage: 1140.471604
</code></pre></div></div>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_1.png"></p>
<p>Memory slowly increases up until over 570 MB, then spikes to 1.14 GB and then drops to 570 MB again, which is the final size of the <code>df_train</code> object. The peak usage being double the final object size may suggest there is an inefficient copy somewhere in the <code>pd.read_csv</code> function, but I’ll let this pass for now.</p>
<p>What should we do next? Train our model! Let us suppose that we’ve already done feature engineering, thus we only need to split our <code>train_data</code> into our design matrix <code>X_train</code> and target <code>y_train</code> and feed them to an <code>ExtraTreesClassifier</code>. Let us check how this step performs:</p>
<div><div><pre><code><span># importing model
</span><span>from</span> <span>sklearn.ensemble</span> <span>import</span> <span>ExtraTreesClassifier</span>

<span>@</span><span>track_memory_use</span><span>(</span><span>close</span><span>=</span><span>False</span><span>,</span> <span>return_history</span><span>=</span><span>True</span><span>)</span>
<span>def</span> <span>fit_model</span><span>(</span><span>train_data</span><span>):</span>
    
    <span># splitting
</span>    <span>X_train</span> <span>=</span> <span>train_data</span><span>.</span><span>drop</span><span>(</span><span>'isFraud'</span><span>,</span> <span>axis</span><span>=</span><span>1</span><span>)</span>
    <span>y_train</span> <span>=</span> <span>train_data</span><span>[</span><span>'isFraud'</span><span>]</span>
    
    <span># fitting model
</span>    <span>et</span> <span>=</span> <span>ExtraTreesClassifier</span><span>(</span><span>n_estimators</span><span>=</span><span>10</span><span>,</span> <span>min_samples_leaf</span><span>=</span><span>10</span><span>)</span>
    <span>et</span><span>.</span><span>fit</span><span>(</span><span>X_train</span><span>,</span> <span>y_train</span><span>)</span>
    
    <span>return</span> <span>et</span>

<span>model</span><span>,</span> <span>mem_history_2</span> <span>=</span> <span>fit_model</span><span>(</span><span>df_train</span><span>)</span>
</code></pre></div></div>
<div><div><pre><code>Current memory usage: 593.745988
Peak memory usage: 1916.650081
</code></pre></div></div>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_2.png"></p>
<p>We jump to a peak of 1.91 GB, and get a steady usage of around 1.5GB. There are little “spikes” throughout the fit, which I suspect are the fitting processes of the individual trees (10 trees = 10 spikes). We end the process with 593 MB.</p>
<p>The final step in our pipeline is reading our test data, performing inference with our model, and reporting the test AUC:</p>
<div><div><pre><code><span># importing metric
</span><span>from</span> <span>sklearn.metrics</span> <span>import</span> <span>roc_auc_score</span>

<span>@</span><span>track_memory_use</span><span>(</span><span>close</span><span>=</span><span>True</span><span>,</span> <span>return_history</span><span>=</span><span>True</span><span>)</span>
<span>def</span> <span>read_test_and_score</span><span>(</span><span>model</span><span>):</span>
    
    <span># reading test
</span>    <span>df_test</span> <span>=</span> <span>pd</span><span>.</span><span>read_csv</span><span>(</span><span>'./test.csv'</span><span>)</span>

    <span># splitting design matrix and target
</span>    <span>X_test</span> <span>=</span> <span>df_test</span><span>.</span><span>drop</span><span>(</span><span>'isFraud'</span><span>,</span> <span>axis</span><span>=</span><span>1</span><span>)</span>
    <span>y_test</span> <span>=</span> <span>df_test</span><span>[</span><span>'isFraud'</span><span>]</span>
    
    <span># scoring and printing result
</span>    <span>preds</span> <span>=</span> <span>model</span><span>.</span><span>predict_proba</span><span>(</span><span>X_test</span><span>)</span>
    <span>score</span> <span>=</span> <span>roc_auc_score</span><span>(</span><span>y_test</span><span>,</span> <span>preds</span><span>[:,</span><span>1</span><span>])</span>
    <span>print</span><span>(</span><span>f</span><span>'Model result is: </span><span>{</span><span>score</span><span>:.</span><span>4</span><span>f</span><span>}</span><span>\n</span><span>'</span><span>)</span>
    
<span>_</span><span>,</span> <span>mem_history_3</span> <span>=</span> <span>read_test_and_score</span><span>(</span><span>model</span><span>)</span>
</code></pre></div></div>
<div><div><pre><code>Model result is: 0.9915

Current memory usage: 596.013196
Peak memory usage: 1069.332149
</code></pre></div></div>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_3.png"></p>
<p>We go from the previous step usage of around 600MB to a peak of 1.1GB and stop at around 600MB again. Also, our model’s AUC is looking great, at <code>0.9915</code>. Putting it all together, our pipeline memory usage goes like this:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_4.png"></p>
<p>This is actually not much, but who knows, tomorrow we may be dealing with datasets and memory usages in the tens or hundreds of GBs. So, let us imagine that we’re limited to less memory than we used in the naive case, for the sake of this post. The same tricks that we’ll devise in this case should work for much larger datasets.</p>
<p>So, what can we do?</p>
<h2>0. Use Dask instead of pandas</h2>
<p>A simple trick to constrain memory usage is replacing <strong>pandas</strong> for <strong>dask</strong>, since it is almost trivial (dask API is a subset of pandas, so it should work as drop-in replacement). Dask can do a lot more than I’ll show here, scaling computation to thousands of machines in a cluster. For our purposes, however, we’ll use it to constrain RAM usage, as it can distribute computation and use the disk for storing intermediate results.</p>
<p>Reading the dataframe using dask yields negligible RAM usage, while keeping the exact pandas syntax (just replace <code>pd</code> with <code>dd</code>):</p>
<div><div><pre><code><span>import</span> <span>dask.dataframe</span> <span>as</span> <span>dd</span>

<span># using a function so we can track memory usage
</span><span>@</span><span>track_memory_use</span><span>(</span><span>plot</span><span>=</span><span>False</span><span>)</span>
<span>def</span> <span>dask_read</span><span>(</span><span>blocksize</span><span>):</span>
    
    <span># reading train data
</span>    <span>df_train</span> <span>=</span> <span>dd</span><span>.</span><span>read_csv</span><span>(</span><span>'./train.csv'</span><span>)</span>

    <span>return</span> <span>dd</span><span>.</span><span>read_csv</span><span>(</span><span>'./train.csv'</span><span>,</span> <span>blocksize</span><span>=</span><span>blocksize</span><span>)</span>

<span># executing
</span><span>df_dd</span> <span>=</span> <span>dask_read</span><span>(</span><span>blocksize</span><span>=</span><span>50e6</span><span>)</span>
</code></pre></div></div>
<div><div><pre><code>Current memory usage: 1.541610
Peak memory usage: 1.897767
</code></pre></div></div>
<p>That’s because dask did not actually read the entire dataset. When we call <code>dd.read_csv</code> we perform two things:</p>
<ol>
<li>Scan the dataframe columns and infer dtypes</li>
<li>Split the dataset into partitions with size <code>blocksize</code> and build the computational graph to process these partitions independently</li>
</ol>
<p>For instance, when we check the object <code>df_dd</code>, it just gives us our dataframe’s structure:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_5.png"></p>
<p>As for the computational graph, we can visualize it by using the <code>.visualize()</code> method:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_6.png"></p>
<p>This graph tells us that dask will independently process eight partitions of our dataframe when we actually do perform computations. This is how distributed computation works: you break your work into smaller, independent pieces and run them in parallel.</p>
<p>But how do we compute stuff using dask? We use the <code>.compute()</code> method. Let us try it for checking how unbalanced our target variable <code>isFraud</code> is, by getting the prevalence of <code>1</code>’s in it.</p>
<p>For a fair comparison to pandas, we will track memory usage all the way from reading the data, and will set <code>scheduler='synchronous'</code> in the <code>.compute()</code> method, so we run single-threaded.</p>
<p>Let us start with pandas. We read the dataframe, calculate the fraction of frauds in the dataset, store it in the variable <code>fraud_prevalence</code>, and finally print the value:</p>
<div><div><pre><code><span>@</span><span>track_memory_use</span><span>()</span>
<span>def</span> <span>pandas_count_target</span><span>():</span>
    <span>df</span> <span>=</span> <span>pd</span><span>.</span><span>read_csv</span><span>(</span><span>'./train.csv'</span><span>)</span>
    <span>fraud_prevalence</span> <span>=</span> <span>(</span><span>df</span><span>[</span><span>'isFraud'</span><span>].</span><span>sum</span><span>()</span><span>/</span><span>df</span><span>[</span><span>'isFraud'</span><span>].</span><span>shape</span><span>[</span><span>0</span><span>])</span>
    <span>print</span><span>(</span><span>f</span><span>'Fraud prevalence is: </span><span>{</span><span>fraud_prevalence</span><span>*</span><span>100</span><span>:.</span><span>3</span><span>f</span><span>}</span><span>%</span><span>\n</span><span>'</span><span>)</span>

<span>pandas_count_target</span><span>()</span>
</code></pre></div></div>
<div><div><pre><code>Fraud prevalence is: 0.128%

Current memory usage: 0.159460
Peak memory usage: 1140.358535
</code></pre></div></div>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_7.png"></p>
<p>As calculating <code>fraud_prevalence</code> is fairly simple, the memory usage is almost the same as reading the dataframe, apart from the fact that we don’t return it and thus get a very low memory footprint when we step out of the function.</p>
<p>Let us now use dask. The syntax is exactly the same, apart from the <code>.compute(scheduler='synchronous')</code> method, which instructs dask to not only build the graph, but actually perform computations.</p>
<div><div><pre><code><span>@</span><span>track_memory_use</span><span>()</span>
<span>def</span> <span>dask_count_target</span><span>(</span><span>blocksize</span><span>):</span>
    <span>df</span> <span>=</span> <span>dd</span><span>.</span><span>read_csv</span><span>(</span><span>'./train.csv'</span><span>,</span> <span>blocksize</span><span>=</span><span>blocksize</span><span>)</span>
    <span>fraud_prevalence</span> <span>=</span> <span>(</span><span>df</span><span>[</span><span>'isFraud'</span><span>].</span><span>sum</span><span>()</span><span>/</span><span>df</span><span>[</span><span>'isFraud'</span><span>].</span><span>shape</span><span>[</span><span>0</span><span>]).</span><span>compute</span><span>(</span><span>scheduler</span><span>=</span><span>'synchronous'</span><span>)</span>
    <span>print</span><span>(</span><span>f</span><span>'Fraud prevalence is: </span><span>{</span><span>fraud_prevalence</span><span>*</span><span>100</span><span>:.</span><span>3</span><span>f</span><span>}</span><span>%</span><span>\n</span><span>'</span><span>)</span>

<span>dask_count_target</span><span>(</span><span>blocksize</span><span>=</span><span>50e6</span><span>)</span>
</code></pre></div></div>
<div><div><pre><code>Fraud prevalence is: 0.128%

Current memory usage: 26.522038
Peak memory usage: 249.080990
</code></pre></div></div>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_8.png"></p>
<p>We get this interesting plot. It takes a little bit more time but memory usage is much lower at around a 250 MB peak. Remember the computational graph, broke into eight partitions? As we used a single thread (<code>scheduler='synchronous'</code>) dask performed the computation sequentially, and as we can see in the graph, there are eight “blocks” through time. If we don’t use the <code>'scheduler='synchronous'</code> parameter, dask will distribute computation across cores and threads:</p>
<div><div><pre><code><span>@</span><span>track_memory_use</span><span>()</span>
<span>def</span> <span>dask_count_target</span><span>(</span><span>blocksize</span><span>):</span>
    <span>df</span> <span>=</span> <span>dd</span><span>.</span><span>read_csv</span><span>(</span><span>'./train.csv'</span><span>,</span> <span>blocksize</span><span>=</span><span>blocksize</span><span>)</span>
    <span>fraud_prevalence</span> <span>=</span> <span>(</span><span>df</span><span>[</span><span>'isFraud'</span><span>].</span><span>sum</span><span>()</span><span>/</span><span>df</span><span>[</span><span>'isFraud'</span><span>].</span><span>shape</span><span>[</span><span>0</span><span>]).</span><span>compute</span><span>()</span>
    <span>print</span><span>(</span><span>f</span><span>'Fraud prevalence is: </span><span>{</span><span>fraud_prevalence</span><span>*</span><span>100</span><span>:.</span><span>3</span><span>f</span><span>}</span><span>%</span><span>\n</span><span>'</span><span>)</span>

<span>dask_count_target</span><span>(</span><span>blocksize</span><span>=</span><span>50e6</span><span>)</span>
</code></pre></div></div>
<div><div><pre><code>Fraud prevalence is: 0.128%

Current memory usage: 0.845965
Peak memory usage: 1453.167118
</code></pre></div></div>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_9.png"></p>
<p>It uses even more memory than pandas (not a lot more, though), but it gets to the result much faster (approximately 1s vs. 4s).</p>
<p>Our simple operation yields a very cool computational graph, shown in the figure below. For clarity, I now use <code>blocksize=100e6</code> which splits the dataset into 4 partitions. For each partition, dask calculates a <code>sum-chunk</code> and a <code>size-chunk</code> which are the sum of the <code>isFraud</code> variable for the partition and the number of rows of the partition, respectively. Then, dask aggregates the <code>sum-chunks</code> and the <code>size-chunks</code> together into <code>sum-agg</code> and <code>size-agg</code>. Finally, dask divides these values to get the prevalence. Even though we depend on scanning all the data to get our result, much of the work can be done in parallel, and dask cleverly makes use of the parallelism and abstracts the details away from us.</p>
<div><div><pre><code><span># let us check the computational graph
</span><span>df</span> <span>=</span> <span>dd</span><span>.</span><span>read_csv</span><span>(</span><span>'./train.csv'</span><span>,</span> <span>blocksize</span><span>=</span><span>100e6</span><span>)</span>
<span>fraud_prevalence</span> <span>=</span> <span>(</span><span>df</span><span>[</span><span>'isFraud'</span><span>].</span><span>sum</span><span>()</span><span>/</span><span>df</span><span>[</span><span>'isFraud'</span><span>].</span><span>shape</span><span>[</span><span>0</span><span>])</span>
<span>graph</span> <span>=</span> <span>fraud_prevalence</span><span>.</span><span>visualize</span><span>()</span>
<span>graph</span>
</code></pre></div></div>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_10.png"></p>
<p>Going back to the sequential scheduler, if we want to further reduce peak memory usage, we can just reduce <code>blocksize</code>:</p>
<div><div><pre><code><span>@</span><span>track_memory_use</span><span>()</span>
<span>def</span> <span>dask_count_target</span><span>(</span><span>blocksize</span><span>):</span>
    <span>df</span> <span>=</span> <span>dd</span><span>.</span><span>read_csv</span><span>(</span><span>'./train.csv'</span><span>,</span> <span>blocksize</span><span>=</span><span>blocksize</span><span>)</span>
    <span>fraud_prevalence</span> <span>=</span> <span>(</span><span>df</span><span>[</span><span>'isFraud'</span><span>].</span><span>sum</span><span>()</span><span>/</span><span>df</span><span>[</span><span>'isFraud'</span><span>].</span><span>shape</span><span>[</span><span>0</span><span>]).</span><span>compute</span><span>(</span><span>scheduler</span><span>=</span><span>'synchronous'</span><span>)</span>
    <span>print</span><span>(</span><span>f</span><span>'Fraud prevalence is: </span><span>{</span><span>fraud_prevalence</span><span>*</span><span>100</span><span>:.</span><span>3</span><span>f</span><span>}</span><span>%</span><span>\n</span><span>'</span><span>)</span>

<span>dask_count_target</span><span>(</span><span>blocksize</span><span>=</span><span>10e6</span><span>)</span>
</code></pre></div></div>
<div><div><pre><code>Fraud prevalence is: 0.128%

Current memory usage: 4.690656
Peak memory usage: 50.915560
</code></pre></div></div>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_11.png"></p>
<p>This further reduces peak memory consumption to 50MB, with the tradeoff that the operation takes 1 second longer as compared to pandas.</p>
<p>Cool! By using dask, we can (almost) still perform computations as we would do in pandas, but with a fairly reduced memory footprint. But our goal is not moving data around, but rather fitting models and solving our fraud analytics problem. Let us then use dask as a platform to improve our whole pipeline, along with some other tricks.</p>
<h3>1. Train with a sample</h3>
<p>Train with a sample? Yeah, I know…</p>
<p>When you’re a data scientist, it’s very hard not to use all the data: it is sitting there, waiting to be analyzed! But, using a sample of your data is actually a very simple and effective way to reduce memory overhead when training models. Remember that we’re using statistical learning: if the number of samples is large enough, adding more data to your model has diminishing returns. So let us test doing this simple trick first!</p>
<p>But we still have to scan the whole dataset to extract a sample of it! That’s why we have dask! Let us rewrite the <code>pandas_read</code> function to a <code>dask_read_and_sample</code> function that reads and samples the data with dask:</p>
<div><div><pre><code><span># using a function so we can track memory usage
</span><span>@</span><span>track_memory_use</span><span>(</span><span>close</span><span>=</span><span>False</span><span>,</span> <span>return_history</span><span>=</span><span>True</span><span>)</span>
<span>def</span> <span>dask_read_and_sample</span><span>(</span><span>blocksize</span><span>,</span> <span>sample_size</span><span>):</span>
    
    <span># reading train data
</span>    <span>df_train</span> <span>=</span> <span>dd</span><span>.</span><span>read_csv</span><span>(</span><span>'./train.csv'</span><span>,</span> <span>blocksize</span><span>=</span><span>blocksize</span><span>)</span>
    
    <span># let us stratify to get the same number of rows for frauds and non-frauds
</span>    <span>sample_positive</span> <span>=</span> <span>df_train</span><span>.</span><span>query</span><span>(</span><span>'isFraud == 1'</span><span>).</span><span>sample</span><span>(</span><span>frac</span><span>=</span><span>sample_size</span><span>)</span>
    <span>sample_negative</span> <span>=</span> <span>df_train</span><span>.</span><span>query</span><span>(</span><span>'isFraud == 0'</span><span>).</span><span>sample</span><span>(</span><span>frac</span><span>=</span><span>sample_size</span><span>)</span>
    
    <span># concatenate the dataframe
</span>    <span>df_sampled</span> <span>=</span> <span>dd</span><span>.</span><span>concat</span><span>([</span><span>sample_positive</span><span>,</span> <span>sample_negative</span><span>])</span>
    
    <span>return</span> <span>df_sampled</span><span>.</span><span>compute</span><span>(</span><span>scheduler</span><span>=</span><span>'synchronous'</span><span>)</span>

<span># executing
</span><span>df_train</span><span>,</span> <span>mem_history_1</span> <span>=</span> <span>dask_read_and_sample</span><span>(</span><span>blocksize</span><span>=</span><span>10e6</span><span>,</span> <span>sample_size</span><span>=</span><span>0.05</span><span>)</span>
</code></pre></div></div>
<div><div><pre><code>Current memory usage: 31.991831
Peak memory usage: 81.060181
</code></pre></div></div>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_12.png"></p>
<p>Cool. We’ve got a great memory reduction with a peak rate of 81 MB and final usage of 32 MB, as expected, as we took a 5% sample of our dataset. Let us check how the rest of the pipeline fares:</p>
<div><div><pre><code><span>model</span><span>,</span> <span>mem_history_2</span> <span>=</span> <span>fit_model</span><span>(</span><span>df_train</span><span>)</span>
<span>_</span><span>,</span> <span>mem_history_3</span> <span>=</span> <span>read_test_and_score</span><span>(</span><span>model</span><span>)</span>
</code></pre></div></div>
<div><div><pre><code>Current memory usage: 78.268949
Peak memory usage: 98.597381

Model result is: 0.9768
Current memory usage: 32.833247
Peak memory usage: 506.189869
</code></pre></div></div>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_13.png"></p>
<p>Cool! We went from a 1.91 GB peak to a 536 MB peak. But there’s some things we now can improve further:</p>
<ol>
<li>Our memory bottleneck now is the scoring part of the pipeline</li>
<li>Our AUC dropped from <code>0.9915</code> to <code>0.9768</code>. This may be acceptable, but it’s kinda hard to let it go…</li>
</ol>
<p>Let us improve our tricks!</p>
<h2>2. Perform inference in blocks</h2>
<p>Now our bottleneck is model inference. As the test dataset is 4x smaller than the train dataset, I did not bother to use dask instead of pandas in the scoring part of the pipeline. Let us change that, and make our model perform inference block by block:</p>
<div><div><pre><code><span>@</span><span>track_memory_use</span><span>(</span><span>close</span><span>=</span><span>True</span><span>,</span> <span>return_history</span><span>=</span><span>True</span><span>)</span>
<span>def</span> <span>dask_read_test_and_score</span><span>(</span><span>model</span><span>,</span> <span>blocksize</span><span>):</span>
    
    <span># reading test
</span>    <span>df_test</span> <span>=</span> <span>dd</span><span>.</span><span>read_csv</span><span>(</span><span>'./test.csv'</span><span>,</span> <span>blocksize</span><span>=</span><span>blocksize</span><span>)</span>

    <span># splitting design matrix and target
</span>    <span>X_test</span> <span>=</span> <span>df_test</span><span>.</span><span>drop</span><span>(</span><span>'isFraud'</span><span>,</span> <span>axis</span><span>=</span><span>1</span><span>)</span>
    <span>y_test</span> <span>=</span> <span>df_test</span><span>[</span><span>'isFraud'</span><span>].</span><span>persist</span><span>(</span><span>scheduler</span><span>=</span><span>'synchronous'</span><span>)</span>
    
    <span># scoring and printing result
</span>    <span>preds</span> <span>=</span> <span>X_test</span><span>.</span><span>map_partitions</span><span>(</span><span>model</span><span>.</span><span>predict_proba</span><span>).</span><span>compute</span><span>(</span><span>scheduler</span><span>=</span><span>'synchronous'</span><span>)</span>
    <span>score</span> <span>=</span> <span>roc_auc_score</span><span>(</span><span>y_test</span><span>,</span> <span>preds</span><span>[:,</span><span>1</span><span>])</span>
    <span>print</span><span>(</span><span>f</span><span>'Model result is: </span><span>{</span><span>score</span><span>:.</span><span>4</span><span>f</span><span>}</span><span>\n</span><span>'</span><span>)</span>
    
<span>_</span><span>,</span> <span>mem_history_3</span> <span>=</span> <span>dask_read_test_and_score</span><span>(</span><span>model</span><span>,</span> <span>blocksize</span><span>=</span><span>5e6</span><span>)</span>
</code></pre></div></div>
<div><div><pre><code>Model result is: 0.9768

Current memory usage: 180.249329
Peak memory usage: 242.521037
</code></pre></div></div>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_14.png"></p>
<p>The implementation is fairly easy. Just use <code>dd.read_csv</code> and set a block size, divide the data into <code>X_test</code> and <code>y_test</code>, and use <code>X_test.map_partitions(model.predict_proba)</code> and <code>.compute(scheduler='synchronous')</code> which will apply our trained model to each partition and get predictions sequentially. As AUC is hard to distribute, I made the choice to bring <code>y_test</code> to memory via the <code>.persist()</code> method and calculate the AUC in-memory as well, but this may be improved after some thought.</p>
<p>Let us check the complete pipeline results:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_15.png"></p>
<p>Much better!!! We’ve reduced peak memory usage by a factor of almost 10x in comparison to the naive model. But it was at the cost of model performance. Can we improve AUC without sacrificing memory usage?</p>
<h2>3. <em>Train with a sample</em>, but on steroids: Bagging</h2>
<p>What if training with a sample was the same as training with the full dataset? Instead of producing just a single model, we can run <code>dask_read_and_sample</code> and <code>fit_model</code> multiple times and combine the models produced in an ensemble. This is akin to Bagging, but with a reduced sample! It will take a lot more time, but it will allow us to stay low on memory and maybe improve our AUC.</p>
<p>We fit 10 models, each on a different 5% sample of the dataset, with replacement:</p>
<div><div><pre><code><span>class</span> <span>EnsembleWrapper</span><span>:</span>
    
    <span>"""
Wrapper for our ensemble.
"""</span>
    
    <span>def</span> <span>__init__</span><span>(</span><span>self</span><span>,</span> <span>model_list</span><span>):</span>
        <span>self</span><span>.</span><span>model_list</span> <span>=</span> <span>model_list</span>
    <span>def</span> <span>predict_proba</span><span>(</span><span>self</span><span>,</span> <span>X</span><span>):</span>
        <span>preds_list</span> <span>=</span> <span>[</span><span>e</span><span>.</span><span>predict_proba</span><span>(</span><span>X</span><span>)</span> <span>for</span> <span>e</span> <span>in</span> <span>self</span><span>.</span><span>model_list</span><span>]</span>
        <span>return</span> <span>np</span><span>.</span><span>array</span><span>(</span><span>preds_list</span><span>).</span><span>mean</span><span>(</span><span>axis</span><span>=</span><span>0</span><span>)</span>

<span># using a function so we can track memory usage
</span><span>@</span><span>track_memory_use</span><span>(</span><span>close</span><span>=</span><span>False</span><span>,</span> <span>return_history</span><span>=</span><span>True</span><span>)</span>
<span>def</span> <span>dask_read_sample_and_fit_model</span><span>(</span><span>blocksize</span><span>,</span> <span>sample_size</span><span>,</span> <span>n_models</span><span>):</span>
    
    <span># list of models
</span>    <span>model_list</span> <span>=</span> <span>[]</span>
    
    <span># loop for each model
</span>    <span>for</span> <span>_</span> <span>in</span> <span>range</span><span>(</span><span>n_models</span><span>):</span>
    
        <span># reading train data
</span>        <span>df_train</span> <span>=</span> <span>dask_read_and_sample</span><span>.</span><span>__wrapped__</span><span>(</span><span>blocksize</span><span>=</span><span>10e6</span><span>,</span> <span>sample_size</span><span>=</span><span>sample_size</span><span>)</span>

        <span># fitting model
</span>        <span>model_list</span><span>.</span><span>append</span><span>(</span><span>fit_model</span><span>.</span><span>__wrapped__</span><span>(</span><span>df_train</span><span>))</span>
    
    <span>return</span> <span>EnsembleWrapper</span><span>(</span><span>model_list</span><span>)</span>
    
<span># executing
</span><span>model</span><span>,</span> <span>mem_history_1</span> <span>=</span> <span>dask_read_sample_and_fit_model</span><span>(</span><span>blocksize</span><span>=</span><span>10e6</span><span>,</span> <span>sample_size</span><span>=</span><span>0.05</span><span>,</span> <span>n_models</span><span>=</span><span>10</span><span>)</span>
<span>_</span><span>,</span> <span>mem_history_2</span> <span>=</span> <span>dask_read_test_and_score</span><span>(</span><span>model</span><span>,</span> <span>blocksize</span><span>=</span><span>5e6</span><span>)</span>
</code></pre></div></div>
<div><div><pre><code>Current memory usage: 2.437998
Peak memory usage: 113.415603

Model result is: 0.9932
Current memory usage: 157.656477
Peak memory usage: 216.875133
</code></pre></div></div>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_16.png"></p>
<p>Our memory use is controlled, our result improved at <code>0.9932</code>, even better than the full model at <code>0.9915</code>. However, we are subject to the tradeoff of the pipeline taking longer to fit (but we could parallelize the ensemble construction).</p>
<h2>4. Use Incremental Learning</h2>
<p>Other way to get a good result with a low memory footprint is using Incremental Learning, which is feeding chunks of data to the model and partially fitting it, one chunk at a time.</p>
<p>Apart from several advances in incremental learning for all kinds of models (I urge you to check <a href="https://scikit-multiflow.github.io/">scikit-multiflow</a>), there is a whole class of models that are incremental by design: neural nets!</p>
<p>So let us implement a Keras model that updates its weights partition by partition. There’s some complexities about normalizing features that I may have resolved poorly, but the core is creating a loop over partition ids, where we call <code>df_train.get_partition(i)</code>, bringing one partition to memory, and then updating the model calling <code>model.fit()</code> inside the loop. This will effectively make the model do a full pass on the training data, but using much less memory.</p>
<div><div><pre><code><span>import</span> <span>tensorflow</span> <span>as</span> <span>tf</span>
<span>from</span> <span>tensorflow.keras</span> <span>import</span> <span>Sequential</span>
<span>from</span> <span>tensorflow.keras.layers</span> <span>import</span> <span>Dense</span><span>,</span> <span>Dropout</span>
<span>tf</span><span>.</span><span>keras</span><span>.</span><span>backend</span><span>.</span><span>set_floatx</span><span>(</span><span>'float64'</span><span>)</span>

<span>class</span> <span>KerasWrapper</span><span>:</span>
    <span>def</span> <span>__init__</span><span>(</span><span>self</span><span>,</span> <span>model</span><span>,</span> <span>feat_mean</span><span>,</span> <span>feat_std</span><span>):</span>
        <span>self</span><span>.</span><span>model</span> <span>=</span> <span>model</span>
        <span>self</span><span>.</span><span>feat_mean</span> <span>=</span> <span>feat_mean</span>
        <span>self</span><span>.</span><span>feat_std</span> <span>=</span> <span>feat_std</span>
        
    <span>def</span> <span>predict_proba</span><span>(</span><span>self</span><span>,</span> <span>X</span><span>):</span>
        
        <span>preds</span> <span>=</span> <span>self</span><span>.</span><span>model</span><span>.</span><span>predict</span><span>((</span><span>X</span> <span>-</span> <span>self</span><span>.</span><span>feat_mean</span><span>)</span><span>/</span><span>self</span><span>.</span><span>feat_std</span><span>)</span>
        <span>return</span> <span>np</span><span>.</span><span>c_</span><span>[</span><span>preds</span><span>,</span> <span>preds</span><span>]</span>
        
<span># using a function so we can track memory usage
</span><span>@</span><span>track_memory_use</span><span>(</span><span>close</span><span>=</span><span>False</span><span>,</span> <span>return_history</span><span>=</span><span>True</span><span>)</span>
<span>def</span> <span>dask_read_and_incrementally_fit_keras</span><span>(</span><span>blocksize</span><span>):</span>
    
    <span># reading df with dask
</span>    <span>df_train</span> <span>=</span> <span>dd</span><span>.</span><span>read_csv</span><span>(</span><span>'./train.csv'</span><span>,</span> <span>blocksize</span><span>=</span><span>blocksize</span><span>)</span>
    
    <span># creating keras model
</span>    <span>model</span> <span>=</span> <span>Sequential</span><span>([</span><span>Dense</span><span>(</span><span>16</span><span>,</span> <span>activation</span><span>=</span><span>'relu'</span><span>),</span>
                        <span>Dropout</span><span>(</span><span>0.25</span><span>),</span>
                        <span>Dense</span><span>(</span><span>16</span><span>,</span> <span>activation</span><span>=</span><span>'relu'</span><span>),</span>
                        <span>Dropout</span><span>(</span><span>0.25</span><span>),</span>
                        <span>Dense</span><span>(</span><span>16</span><span>,</span> <span>activation</span><span>=</span><span>'relu'</span><span>),</span>
                        <span>Dropout</span><span>(</span><span>0.25</span><span>),</span>
                        <span>Dense</span><span>(</span><span>1</span><span>,</span> <span>activation</span><span>=</span><span>'sigmoid'</span><span>)])</span>
    <span>model</span><span>.</span><span>compile</span><span>(</span><span>loss</span><span>=</span><span>'binary_crossentropy'</span><span>)</span>
    
    <span># getting mean and std for dataset to normalize features
</span>    <span>feat_mean</span> <span>=</span> <span>df_train</span><span>.</span><span>drop</span><span>(</span><span>'isFraud'</span><span>,</span> <span>axis</span><span>=</span><span>1</span><span>).</span><span>mean</span><span>().</span><span>compute</span><span>(</span><span>scheduler</span><span>=</span><span>'synchronous'</span><span>)</span>
    <span>feat_std</span> <span>=</span> <span>df_train</span><span>.</span><span>drop</span><span>(</span><span>'isFraud'</span><span>,</span> <span>axis</span><span>=</span><span>1</span><span>).</span><span>std</span><span>().</span><span>compute</span><span>(</span><span>scheduler</span><span>=</span><span>'synchronous'</span><span>)</span>
    
    <span># loop for number of partitions
</span>    <span>for</span> <span>i</span> <span>in</span> <span>range</span><span>(</span><span>df_train</span><span>.</span><span>npartitions</span><span>):</span>
        
        <span># getting one partition
</span>        <span>part</span> <span>=</span> <span>df_train</span><span>.</span><span>get_partition</span><span>(</span><span>i</span><span>).</span><span>compute</span><span>(</span><span>scheduler</span><span>=</span><span>'synchronous'</span><span>)</span>
        
        <span># splitting
</span>        <span>X_part</span> <span>=</span> <span>(</span><span>part</span><span>.</span><span>drop</span><span>(</span><span>'isFraud'</span><span>,</span> <span>axis</span><span>=</span><span>1</span><span>)</span> <span>-</span> <span>feat_mean</span><span>)</span><span>/</span><span>feat_std</span>
        <span>y_part</span> <span>=</span> <span>part</span><span>[</span><span>'isFraud'</span><span>]</span>
        
        <span># running partial fit
</span>        <span>model</span><span>.</span><span>fit</span><span>(</span><span>X_part</span><span>,</span> <span>y_part</span><span>,</span> <span>batch_size</span><span>=</span><span>512</span><span>)</span>
    
    <span>return</span> <span>KerasWrapper</span><span>(</span><span>model</span><span>,</span> <span>feat_mean</span><span>,</span> <span>feat_std</span><span>)</span>

<span>model</span><span>,</span> <span>mem_history_1</span> <span>=</span> <span>dask_read_and_incrementally_fit_keras</span><span>(</span><span>blocksize</span><span>=</span><span>5e6</span><span>)</span>
</code></pre></div></div>
<div><div><pre><code>130/130 [==============================] - 0s 1ms/step - loss: 0.4012
130/130 [==============================] - 0s 1ms/step - loss: 0.0321
130/130 [==============================] - 0s 1ms/step - loss: 0.0185
130/130 [==============================] - 0s 1ms/step - loss: 0.0143
130/130 [==============================] - 0s 1ms/step - loss: 0.0102
130/130 [==============================] - 0s 1ms/step - loss: 0.0100
130/130 [==============================] - 0s 1ms/step - loss: 0.0087
130/130 [==============================] - 0s 1ms/step - loss: 0.0090
130/130 [==============================] - 0s 1ms/step - loss: 0.0086
130/130 [==============================] - 0s 1ms/step - loss: 0.0079
130/130 [==============================] - 0s 1ms/step - loss: 0.0072
...

Current memory usage: 114.445448
Peak memory usage: 161.811473
</code></pre></div></div>
<p>Inference with a neural net seems a little bit more expensive in terms of memory:</p>
<div><div><pre><code><span>_</span><span>,</span> <span>mem_history_2</span> <span>=</span> <span>dask_read_test_and_score</span><span>(</span><span>model</span><span>,</span> <span>blocksize</span><span>=</span><span>5e6</span><span>)</span>
</code></pre></div></div>
<div><div><pre><code>Model result is: 0.9833

Current memory usage: 318.801547
Peak memory usage: 358.292797
</code></pre></div></div>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_17.png"></p>
<p>We get an AUC of <code>0.9833</code>, around 45s of runtime, and 360 MB of peak memory. It seems like a common ground between memory usage, performance, and execution time.</p>
<h2>Final comparison</h2>
<p>I shouldn’t leave you without a plot with a final comparison of our tricks:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/big_data_ml_training/bdml_18.png"></p>
<p>I’ll end the post here, but there’s a lot more tricks to try:</p>
<ul>
<li>Test different sample sizes (we were very agressive with 5%)</li>
<li>Run a feature selection method on a sample, drop unused features on the full dataframe and fit model (using dask). This should save orders of magnitude of memory on high dimensional datasets.</li>
<li>Solve pandas “spike” problem. Remember that pandas had a peak meamory use of double the size of the dataframe? This also deteriorates dask performance as it is running the same pandas over its partitions and showing the same spiky behavior.</li>
<li>Use Dask-ML. It is build on top of <code>sklearn</code> and shares a common interface with it, same as <code>dask.dataframe</code> and pandas. It has the “incremental scoring” capability ready for sklearn models.</li>
<li>many more that I’ll remember eventually!</li>
</ul>
<p>I hope this post helps you! Full code available <a href="https://github.com/gdmarmerola/big-data-ml-training/tree/master">here</a>.</p></article></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Guilherme’s Blog <br>
<br>
February 5, 2024 at 04:43PM
